topo/node/ceos/ceos.go

```diff
@@ -83,127 +149,95 @@ func (n *Node) GenerateSelfSigned(ctx context.Context, ni node.Interface) error
 		}
 	}
 	log.Infof("%s - pod running.", n.pb.Name)
-	done := false
-	var g expect.Expecter
-	log.Infof("%s - waiting on container to be ready", n.pb.Name)
-	waitTime := 0 * timeSecond
-	for !done {
-		time.Sleep(waitTime)
-		cmd := fmt.Sprintf("kubectl exec -it -n %s %s -- Cli", ni.Namespace(), n.pb.Name)
-		var err error
-		g, _, err = spawner(cmd, -1)
-		// This could be set per error case but probably not worth it.
-		waitTime = 10 * timeSecond
-		if err != nil {
-			log.Debugf("%s - process not ready - waiting.", n.pb.Name)
-			continue
-		}
-		_, _, err = g.Expect(regexp.MustCompile(`>`), -1)
-		if err != nil {
-			log.Debugf("%s - os not ready - waiting.", n.pb.Name)
-			continue
-		}
-		log.Debugf("%s - captured prompt", n.pb.Name)
-		if err := g.Send("enable\n"); err != nil {
-			return err
-		}
-		_, _, err = g.Expect(regexp.MustCompile(`#`), 5*timeSecond)
-		if err != nil {
-			log.Debugf("%s - auth not ready - waiting.", n.pb.Name)
-			continue
-		}
-		done = true
-	}
-	cmd := fmt.Sprintf("security pki key generate rsa %d %s\n", selfSigned.KeySize, selfSigned.KeyName)
-	log.Debugf("%s - enabled - sending %q", n.pb.Name, cmd)
-	if err := g.Send(cmd); err != nil {
-		return err
-	}
-	_, _, err = g.Expect(regexp.MustCompile(`#`), 30*timeSecond)
+
+	err = n.SpawnCLIConn(ni.Namespace())
 	if err != nil {
 		return err
 	}
-	cmd = fmt.Sprintf("security pki certificate generate self-signed %s key %s parameters common-name %s\n", selfSigned.CertName, selfSigned.KeyName, n.pb.Name)
-	log.Debugf("key generated - sending %q", cmd)
-	if err := g.Send(cmd); err != nil {
-		return err
+
+	defer n.cliConn.Close()
+
+	cfgs := []string{
+		fmt.Sprintf(
+			"security pki key generate rsa %d %s\n",
+			selfSigned.KeySize,
+			selfSigned.KeyName,
+		),
+		fmt.Sprintf(
+			"security pki certificate generate self-signed %s key %s parameters common-name %s\n",
+			selfSigned.CertName,
+			selfSigned.KeyName,
+			n.pb.Name,
+		),
 	}
-	_, _, err = g.Expect(regexp.MustCompile(`(.*)#`), 30*timeSecond)
+
+	r, err := n.cliConn.SendConfigs(cfgs)
 	if err != nil {
 		return err
+	} else if r.Failed() {
```

Comment:
Also i would question why is the this the API pattern?

Either "sendconfigs" works it doesn't

if you want to have a different action based on the type of error then build that into the api

https://play.golang.org/p/P9LcmbJna6M

basically the go style would say if you get an err you cannot trust the return value.

there are a couple examples in go there that is done - for one the file io functions do return both return values and err's specifically for handling io.EOF but that is really a pattern based on the OS semantics not on it being a good patten

Reply:
Hi @marcushines 
`else if` addressed in eca3d33

the decision to have `r.Failed()` for multiline commands and `r.Failed` for single commands was deliberate. It indeed might look weird, but that is because it is not an error in the context of the `SendXXX` operation.

The way scrapligo treats operations over SSH/TTY transports and its errors is the following:

1. when a user attempts to send a CLI command towards a device it uses multiple vehicles exposed in scrapligo api: `channel`, `session`, `transport` layers .
2. any error that might appear on the above  layers will return an error if things went utterly bad and the attempt to **send** the command failed.
3. if the low level things worked out and the user was able to send the command, scrapligo marks this as a successful operation and the error is `nil`.
4. If a user sends a command such as `**^%*&%` and scrapli detects that this command was received by a device and a device returned expected prompt back, this is a successful operation.
5. Since the act of sending the command/config was successful, we have a `nil` error. It is very much like in HTTP libraries you may succeed in sending HTTP request but you check the status (i.e. for `404`).
6. And that _Status_ is what `Failed` is. It will simply indicate if scrapli detected that whatever the user sent over was not accepted by the Network OS (still the send operation was totally good) application wise.

I hope this makes sense :)

Reply:
Maybe a practical use cases that I just had with ceos that explains it even more precise.

I had the following set of set configs:

```
cfgs := []string{
		"no ip address",
		"no ipv6 address",
		"interface management 0",
	}
```

I sent this with scrapligo `SendConfigs` and what scrapligo did is happily sent it over, declared things were good.
They were good, because scrapligo sent the commands over, they were processed by the CLI and on the transport side of things it all checked out. The commands were succesfully sent and they were applied in the correct privilege level.

The problem is that CLI of EOS emitted some errors, because my order of operations was incorrect. To indicate that there was an application layer issue, scrapligo uses `Failed` flag on the response object.